### ISSUE TITLE
Errors prepare environment failed at cannot connect to Docker daemon

### ISSUE THREAD

**Issue by Nightbr** [cattle-ops/terraform-aws-gitlab-runner#1025]
Hello there, first of all, thank you for this amazing project that helps us a lot to manage our runners fleet.

We use the latest version of the package (7.1.1), gitlab-runner (16.5.0).

# Describe the bug

It happens for around 10-20% of jobs but we get this error:

`ERROR: Job failed (system failure): prepare environment: Cannot connect to the Docker daemon at tcp://10.1.2.59:2376. Is the docker daemon running? (docker.go:570:120s). Check https://docs.gitlab.com/runner/shells/index.html#shell-profile-loading for more information`

![image](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/4228646/1c42979c-4964-4068-a9b8-154008907e5b)

# To Reproduce

Here is my full config

```
module "label_gitlab_runner" {
  source = "../../common/label"
  stage  = var.stage
  name   = "gitlab-runner"
}

data "aws_availability_zones" "available" {
  state = "available"
}

module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "5.1.2"

  name = module.label_gitlab_runner.id
  cidr = var.cidr_block

  azs             = ["${var.region}b"]
  public_subnets  = [cidrsubnet(var.cidr_block, 8, 1)]
  private_subnets = [cidrsubnet(var.cidr_block, 8, 2)]

  # Single NAT Gateway
  enable_nat_gateway     = true
  single_nat_gateway     = true
  one_nat_gateway_per_az = false

  tags = module.label_gitlab_runner.tags
}

module "vpc_endpoints" {
  source  = "terraform-aws-modules/vpc/aws//modules/vpc-endpoints"
  version = "5.1.2"

  vpc_id = module.vpc.vpc_id

  endpoints = {
    s3 = {
      service = "s3"
      tags    = { Name = "s3-vpc-endpoint" }
    }
  }

  tags = module.label_gitlab_runner.tags
}

data "aws_security_group" "default" {
  name   = "default"
  vpc_id = module.vpc.vpc_id
}

# https://github.com/cattle-ops/terraform-aws-gitlab-runner
module "gitlab-runner" {
  source      = "cattle-ops/gitlab-runner/aws"
  version     = "7.1.1"
  environment = var.stage

  vpc_id    = module.vpc.vpc_id
  subnet_id = element(module.vpc.private_subnets, 0)


  runner_instance = {
    name                        = module.label_gitlab_runner.id
    type                        = "t3.small"
    ssm_access                  = true
    collect_autoscaling_metrics = ["GroupDesiredCapacity", "GroupInServiceCapacity"]
  }

  runner_manager = {
    maximum_concurrent_jobs = 50
  }

  runner_gitlab = {
    url = var.runners_gitlab_url
    # Gitlab runner version, change this to update
    # https://gitlab.com/gitlab-org/gitlab-runner/-/releases
    runner_version = "16.5.0"
  }

  # Gitlab runner configuration
  runner_gitlab_registration_config = {
    registration_token = var.runner_token
    tag_list           = "docker, aws, spot_runner"
    description        = "Gitlab runner on AWS Spot"
    locked_to_project  = "false"
    run_untagged       = "true"
    maximum_timeout    = "3600"
  }

  runner_worker = {
    type       = "docker+machine"
    ssm_access = true
    # Fix docker bugs
    environment_variables = [
      "DOCKER_TLS_CERTDIR="
    ]
  }

  runner_worker_docker_machine_fleet = {
    enable = true
  }

  # Options added to the [runners.docker] section of config.toml to configure the Docker container of the Runner Worker.
  # For details check https://docs.gitlab.com/runner/configuration/advanced-configuration.html
  runner_worker_docker_options = {
    wait_for_services_timeout = 120 # Increase to 2min for default of 30s
  }

  runner_worker_docker_machine_instance = {
    # /!\ Note that there is special limit for T3 & T2 instances
    # @doc https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-spot-limits.html#t2-spot-instances
    types      = ["m5d.2xlarge", "c5.2xlarge"]
    subnet_ids = module.vpc.private_subnets
    # @doc https://docs.gitlab.com/runner/configuration/autoscale.html#autoscaling-algorithm-and-parameters
    # Autoscaling parameters off peak.
    idle_count      = 3
    idle_time       = 900 # 15min
    max_growth_rate = 50
    max_builds      = 100
    # The size of the root volume for the Runner Worker
    root_size = 64
  }

  # Autoscaling parameters off peak
  # https://docs.gitlab.com/runner/configuration/advanced-configuration.html#the-runnersmachineautoscaling-sections
  runner_worker_docker_machine_autoscaling_options = [
    {
      periods    = ["* * 0-9,18-23 * * mon-fri *"]
      idle_count = 0
      idle_time  = 600 # 10min
      timezone   = "Europe/Paris"
    },
    {
      periods    = ["* * * * * sat,sun *"]
      idle_count = 0
      idle_time  = 60 # 1min
      timezone   = "Europe/Paris"
    }
  ]

  runner_worker_docker_machine_role = {
    additional_tags = {
      "module"                                 = module.label_gitlab_runner.id
      "tf-aws-gitlab-runner:instancelifecycle" = "spot:yes"
      "VantaNonProd"                           = "true"
    }
  }

  runner_worker_docker_machine_instance_spot = {
    enable = true
    # See pricing History for Spot instances
    # @doc https://docs.aws.amazon.com/fr_fr/AWSEC2/latest/UserGuide/using-spot-instances-history.html
    max_price = "on-demand-price"
  }

  runner_networking = {
    allow_incoming_ping                    = true
    allow_incoming_ping_security_group_ids = [data.aws_security_group.default.id]
    security_group_ids                     = [data.aws_security_group.default.id]
  }

  # Enable automatic redeployment of the Runner ASG when the Launch Configs change.
  runner_enable_asg_recreation = true

  runner_cloudwatch = {
    enable         = true
    log_group_name = "/${module.label_gitlab_runner.id}"
    retention_days = 365
  }

  # For debug purpose
  debug = {
    trace_runner_user_data         = false
    write_runner_config_to_file    = false
    write_runner_user_data_to_file = false
  }

  tags = module.label_gitlab_runner.tags

}


resource "null_resource" "cancel_spot_requests" {
  # Cancel active and open spot requests, terminate instances
  triggers = {
    module = module.label_gitlab_runner.id
  }

  provisioner "local-exec" {
    when    = destroy
    command = "./scripts/cancel-spot-instances.sh ${self.triggers.module}"
  }
}

```

# Expected behavior

No failed job with the error.

# Additional context

I tried different things:

- Upgrade all terraform modules (VPC to 5.1.2, module to latest, AWS to latest too)... -> still the error happening randomly
- The error was always happened during off peak period, so I thought we need to increase the idle_time to let the time for docker to be ready before the autoscaler kick the machine -> not really helpful, still randomly get the error

Any helps or lead would be really appreciated, thanks in advance for your help :pray:

**Comment by kayman-mk** [comment]
Puuh, seems to be a hard one. I have never seen this error before. Some thoughts I had looking at your source:
- the `null_resource` doesn't work as expected and can be removed (see #426). As far as I remember it was introduced to support the examples. 
- what is this Docker fix good for `DOCKER_TLS_CERTDIR=`?

Unless there is a bug in the GitLabRunner implementation, the changing the idle time shouldn't make any difference. You observed the problem while running a job, so the machine is not idle and shouldn't be removed. I use a zero idle time for off peak and have never seen this error.

In case you are not able to connect to the Docker daemon during a shutdown, it is strange that your container is still running. Shutting down the Docker daemon includes killing all containers, so they wouldn't run any longer.

Sorry, no helpful idea at the moment. Playing with the idle time is something you should try.

**Comment by Nightbr** [comment]
Hey,
 
- I have remove the null_resource, still works as expected!
- `DOCKER_TLS_CERTDIR=` is a fix to force Docker to use non TLS connection. I have refactor the config to be able to use Docker TLS connection (which is the default now) and update my .gitlab-ci.yml to make it work

Changes on the main.tf
```

runner_worker = {
    type       = "docker+machine"
    ssm_access = true
  }

  runner_worker_docker_machine_fleet = {
    enable = true
  }

  # Options added to the [runners.docker] section of config.toml to configure the Docker container of the Runner Worker.
  # For details check https://docs.gitlab.com/runner/configuration/advanced-configuration.html
  runner_worker_docker_options = {
    wait_for_services_timeout = 120 # Increase to 2min for default of 30s
    privileged                = "true"
    volumes                   = ["/cache", "/certs/client"]
  }
```

And in .gitlab-ci.yml

```
variables:
  # Docker variables
  # @see https://docs.gitlab.com/ee/ci/docker/using_docker_build.html#use-docker-in-docker
  DOCKER_DRIVER: overlay2
  DOCKER_HOST: 'tcp://docker:2376'
  DOCKER_TLS_CERTDIR: '/certs'
  DOCKER_TLS_VERIFY: '1'
  DOCKER_CERT_PATH: '$DOCKER_TLS_CERTDIR/client'
```

Even with this, we still have some jobs that failed with the same error:

![image](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/4228646/09f19d69-813c-4e9c-a1da-092bcf1b7b7f)

**Comment by kayman-mk** [comment]
Hm, just thinking about my config. I set no `DOCKER_` variables at all and using docker+machine driver (communicates via port 2376, checked with `docker-machine ls` on the Runner). There is no `services: - docker:24.0.5-dind` in my .gitlab-ci.yml and I am using Docker in my jobs. The only thing I found was

```hcl
  runner_worker_docker_options = {
    volumes      = ["/cache", "/var/run/docker.sock:/var/run/docker.sock"]
  }
```

I added the docker executable to my image used in the jobs and I am able to connect to the Docker daemon without any problems. It might be the case that the Docker communication on the Runner Worker is not using TLS. But this is fine as the machine is under my control, communicates with localhost only and does not execute multiple jobs at the same time.
